### ISSUE TITLE
403 error `_xsrf` missing from POST

### ISSUE THREAD

**Issue by SimonBiggs** [ReviewNB/jupyterlab-gitplus#3]
Hi Amit,

Love your work :). Just wanted to give a heads up. Currently trying to set up GitPlus within our environment here. 

I ran the following steps as per the readme:

```
pip install --upgrade jupyterlab_gitplus
jupyter labextension install @reviewnb/jupyterlab_gitplus
jupyter serverextension enable --py jupyterlab_gitplus
```

as well as updating the config file as detailed below:

![image](https://user-images.githubusercontent.com/6559099/79177436-244c1d80-7e46-11ea-94d3-ebfc8ec8d1ce.png)

The error message within the jupyterlab server terminal is the following:

```
[W 01:44:05.261 LabApp] 403 POST /gitplus/modified_repo (127.0.0.1): '_xsrf' argument missing from POST                                                               [W 01:44:05.323 LabApp] 403 POST /gitplus/modified_repo (127.0.0.1) 63.00ms referer=http://lab.physics-server/lab?   
```

I'm more than happy to provide more debugging info. Thanks for building this.

Cheers,
Simon

**Comment by SimonBiggs** [comment]
If it helps I have also had similar issues with an extension I wrote a while ago:

https://github.com/SimonBiggs/scriptedforms/issues/101
https://github.com/SimonBiggs/scriptedforms/issues/57

**Comment by amit1rrr** [comment]
> Love your work :).

Thanks @SimonBiggs :)

> '_xsrf' argument missing from POST    

Reading a few answers on stakcoverflow it seems like this happens when Jupyter notebook browser tab is idle for too long. Can you try refreshing browser tab? If that fails, try restarting Jupyter once just to be sure. Also, when this error happens with GitPlus can you do normal Jupyter server operation such as saving a notebook?


Another possibility is - You have 3 local_hostnames in the config. Is it possible that the browser tab is using one hostname and server APIs are on a different hostname? For this, can you try accessing Jupyter on the same endpoint that is printed on console when you start Jupyter.


These above experiments would help me understand the issue better. Thanks.

**Comment by SimonBiggs** [comment]
To add to that, I had the same issue on a completely different computer (still Windows 10) just using the default localhost host name and configuration. I'll see if the issue is repeatable on my home Linux machine.

**Comment by SimonBiggs** [comment]
And unfortunately the idle tab isn't the solution I opened it immediately after boot, and also tried incognito mode. All had the same issue.

**Comment by amit1rrr** [comment]
I just came across [this article](https://blog.jupyter.org/security-release-jupyter-notebook-4-3-1-808e1f3bb5e2).  It mentions,

> The use of the xsrf token affects existing extensions / etc. that work directly with the notebook server via REST API. To make requests to the rest API, the _xsrf cookie must be sent back to the server in the X-XSRF-Token header of API requests

Just saw that [JupyterLab is also adding _xsrf header](https://github.com/jupyterlab/jupyterlab/blob/6111f2bd15bc8f65f18d78bcc88f8c4d3243f7cc/packages/services/src/serverconnection.ts#L267). 

I will figure out a way to incorporate this in GitPlus and release an update. Stay tuned.

Sorry for the inconvenience.  This issue went undetected because it never occurred in my testing (maybe because I am using the same endpoint 127.0.0.1 everywhere or my xsrf config is disabled.)

**Comment by amit1rrr** [comment]
Meanwhile you can try out the extension by [disabling xsrf](https://github.com/nteract/hydrogen/issues/922#issuecomment-315665849). I wouldn't recommend it as a solution but just temporary workaround until I release a proper fix.

**Comment by SimonBiggs** [comment]
No inconvenience, no need to be sorry :). Glad the tool is being made, will look forward to rolling it out at our centre at work.

It meets quite a need, I have a centre wide hosted JupyterLab server, which is also under got version control. Helping people to create pull requests within JupyterLab as you are enabling it will massively help in the adoption of using got version as the standard.

**Comment by amit1rrr** [comment]
@SimonBiggs The fix for CORS issue is now released in version `0.1.3`

Please upgrade your extension with,
`jupyter labextension install @reviewnb/jupyterlab_gitplus`

You can verify the version with,
`jupyter labextension list`

Feel free to reopen the issue if you face any problems.
